01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+  single space, or return many spaces + versions.
+
+  Technically the `git` resource may push many commits, so this is necessary to
+  track them all as outputs of a build. This could also support batch creation.
+
+  To ensure `check` is the source of truth for ordering, the versions returned
+  by `put` are not order dependent. A `check` will be performed to discover
+  them in the correct order, and then each version will be saved as an output
+  of the build. The latest version of the set will then be fetched.
+
+* Change `put` to write its JSON response to a specified file, rather than
+  `stdout`, so that we don't have to be attached to process its response.
+
+  This is one of the few ways a build can error after the ATC reattaches
+  (`unexpected end of JSON`). With it written to a file, we can just try to read
+  the file when we re-attach after seeing that the process exited. This also
+  frees up stdout/stderr for normal logging, which has been an occasional
+  pitfall during resource development/debugging.
+
+
+## Introduction of "notifier" resource interface
+
+A resource may now implement a "notifier" interface. Notifications must be
+explicitly enabled in the pipeline. This is so that a resource type can't
+suddenly decide to include a notifier and start doing surprising things.
+
+A notifier has a single hook, `notify`, and receives some sort of JSON payload
+describing the event. The event is purely for side-effects and has no
+implications for pipeline semantics.
+
+If a notification fails to send, the build shall error. To be honest, I haven't
+thought about this much yet, but I think it's better to be conservative.
+
+A resource type may implement both "artifacts" and "notifier" interfaces. If a
+resource implements both, it will be given information about its occurrence in
+the build (space and version). This will be crucial for e.g. reporting the
+status of a pull request or commit back to GitHub.
+
+Add support for a "notifier" resource type. They have one hook, `notify`,
+which is invoked with various types of notifications
+(`build_started`, `build_finished`, as the poster children).
+
+
+# Caveats
+
+Terminology is now even more confusing. "Resource type" could mean either the
+"git" vs. "s3" or "notifier" vs. "artifact".
+
+
+# Open Questions
+
+## Can we reduce the `check` overhead?
+
+With spaces there will be more `check`s than ever. Right now, there's one
+container per recurring `check`. Can we reduce the container overhead here by
+requiring that resource `check`s be side-effect free and able to run in parallel?
+
+There may be substantial security implications for this.
+
+
+## Is `destroy` general enough to be a part of the interface?
+
+It may be the case that most resources cannot easily support `destroy`. One
+example is the `git` resource. It doesn't really make sense to `destroy` a
+commit. Even if it did (`push -f`?), it's a kind of weird workflow to support
```

Comment:
for git, yes, it a weird workflow. does every resource need to implement all parts of the interface? Implementing noops is far less cumbersome for resource implementers than adding additional complexity to any one of the interfaces. If info could declare which bits of interface are getting implemented, and then the interaction between all of them could be described, that'd be even better.

Reply:
> I think it's a smell if there are parts of a general interface that are impossible to satisfy with a substantial portion of the implementations

I totally agree and I think we're probably smelling the same inadequacy. if it didn't sound super absurd on its face, I'd think that maybe what we need is something like an antiartifact-- 

it's got `anticheck` and `delete`, but obviously no get. both emit deleted versions in chronological order?

Reply:
Funny that you mention `anticheck` - sounds similar to the ancestry check I just commented here: https://github.com/concourse/rfcs/pull/1#discussion_r182265596